components/mpas-ocean/src/shared/mpas_ocn_vmix.F

```diff
@@ -288,54 +296,67 @@ subroutine ocn_vel_vmix_tend_implicit_rayleigh(meshPool, dt, kineticEnergyCell,
          do k = Nsurf, N
             layerThicknessEdge(k,iEdge) = 0.5_RKIND * (layerThickness(k,cell1) + layerThickness(k,cell2))
             edgeThicknessTotal = edgeThicknessTotal + layerThicknessEdge(k,iEdge)
-         end do
-
-         ! A is lower diagonal term
-         A(1:Nsurf)=0.0_RKIND
-         do k = Nsurf+1, N
-            A(k) = -2.0_RKIND*dt*vertViscTopOfEdge(k,iEdge) &
-               / (layerThicknessEdge(k-1,iEdge) + layerThicknessEdge(k,iEdge)) &
-               / layerThicknessEdge(k,iEdge)
          enddo
 
-         ! C is upper diagonal term
-         C(1:Nsurf-1)=0.0_RKIND
-         do k = Nsurf, N-1
-            C(k) = -2.0_RKIND*dt*vertViscTopOfEdge(k+1,iEdge) &
-               / (layerThicknessEdge(k,iEdge) + layerThicknessEdge(k+1,iEdge)) &
-               / layerThicknessEdge(k,iEdge)
-         enddo
+         ! tridiagonal matrix algorithm
+         C(Nsurf) = -2.0_RKIND*dt*vertViscTopOfEdge(Nsurf+1,iEdge) &
+                / (layerThicknessEdge(Nsurf,iEdge) + layerThicknessEdge(Nsurf+1,iEdge)) &
+                / layerThicknessEdge(Nsurf,iEdge)
+         bTemp(Nsurf) = 1.0_RKIND - C(Nsurf) &
+            ! added Rayleigh terms
+            + dt*rayleighDampingCoef/((1.0_RKIND - rayleighDepthVariable) + rayleighDepthVariable*edgeThicknessTotal)
+         rTemp(Nsurf) = normalVelocity(Nsurf,iEdge)
 
-         ! B is diagonal term
-         B(1:Nsurf-1)=0.0_RKIND
-         B(Nsurf) = 1.0_RKIND - C(Nsurf) &
-           + dt*rayleighDampingCoef/((1.0_RKIND - rayleighDepthVariable) + rayleighDepthVariable*edgeThicknessTotal)
+         ! first pass: set the coefficients
          do k = Nsurf+1, N-1
-            B(k) = 1.0_RKIND - A(k) - C(k) &
-              ! added Rayleigh terms
-              + dt*(rayleighDampingCoef/((1.0_RKIND - rayleighDepthVariable) + rayleighDepthVariable*edgeThicknessTotal))
+            A        = -2.0_RKIND*dt*vertViscTopOfEdge(k,iEdge) &
+                   / (layerThicknessEdge(k-1,iEdge) + layerThicknessEdge(k,iEdge)) &
+                   / layerThicknessEdge(k,iEdge)
+            m        = A/bTemp(k-1)
+            C(k)     = -2.0_RKIND*dt*vertViscTopOfEdge(k+1,iEdge) &
+                   / (layerThicknessEdge(k,iEdge) + layerThicknessEdge(k+1,iEdge)) &
+                   / layerThicknessEdge(k,iEdge)
+            bTemp(k) = 1.0_RKIND - A - C(k) &
+                   ! added Rayleigh terms
+                   + dt*(rayleighDampingCoef/((1.0_RKIND - rayleighDepthVariable) + rayleighDepthVariable*edgeThicknessTotal)) &
+                   - m*C(k-1)
+            rTemp(k) = normalVelocity(k,iEdge) - m*rTemp(k-1)
```

Comment:
@xylar and @philipwjones, the tridiagonal solve is in-lined here. There are changes in order of operations, where BFB between before/after could be compiler and machine dependent, and particularly sensitive to optimization. It would be good to see the order of magnitude of differences with COMPASS.  The standard comparison reports that. If it only loses a few digits from machine precision, I think we can bless the COMPASS diffs. 

I think that all terms in this routine are nonzero in both COMPASS and E3SM (bottom drag, surface forcing, viscosity) so this should not be a difference in test configuration. If anything, E3SM should be more challenging because of time-varying forcing and surface fluxes.

Reply:
@mark-petersen, as I stated above, the comparison is here:
https://github.com/MPAS-Dev/compass/pull/295#issuecomment-991949465

Reply:
The fact that E3SM tests were all bit-for-bit and Compass tests were not suggests to me, as I said, that it just has to be something like differences in compiler flag or something like that.  I just don't see any other way to understand being bit-for-bit in E3SM but not in Compass with the same compilers.